balancer/rls/internal/control_channel.go

```diff
@@ -0,0 +1,222 @@
+	// information contained in headers set by the application.
+	dopts := []grpc.DialOption{grpc.WithAuthority(cc.buildOpts.Authority)}
+	if cc.buildOpts.Dialer != nil {
+		dopts = append(dopts, grpc.WithContextDialer(cc.buildOpts.Dialer))
+	}
+
+	// The control channel will use the channel credentials from the parent
+	// channel, including any call creds associated with the channel creds.
+	var credsOpt grpc.DialOption
+	switch {
+	case cc.buildOpts.DialCreds != nil:
+		credsOpt = grpc.WithTransportCredentials(cc.buildOpts.DialCreds.Clone())
+	case cc.buildOpts.CredsBundle != nil:
+		// The "fallback" mode in google default credentials (which is the only
+		// type of credentials we expect to be used with RLS) uses TLS/ALTS
+		// creds for transport and uses the same call creds as that on the
+		// parent bundle.
+		bundle, err := cc.buildOpts.CredsBundle.NewWithMode(internal.CredsBundleModeFallback)
+		if err != nil {
+			return nil, err
+		}
+		credsOpt = grpc.WithCredentialsBundle(bundle)
+	default:
+		cc.logger.Warningf("no credentials available, using Insecure")
+		credsOpt = grpc.WithInsecure()
+	}
+	return append(dopts, credsOpt), nil
+}
+
+func (cc *controlChannel) monitorConnectivityState() {
+	cc.logger.Infof("Starting connectivity state monitoring goroutine")
+	// Since we use two mechanisms to deal with RLS server being down:
+	//   - adaptive throttling for the channel as a whole
+	//   - exponential backoff on a per-request basis
+	// we need a way to avoid double-penalizing requests by counting failures
+	// toward both mechanisms when the RLS server is unreachable.
+	//
+	// To accomplish this, we monitor the state of the control plane channel. If
+	// the state has been TRANSIENT_FAILURE since the last time it was in state
+	// READY, and it then transitions into state READY, we push on a channel
+	// which is being read by the LB policy.
+	//
+	// The LB the policy will iterate through the cache to reset the backoff
+	// timeouts in all cache entries. Specifically, this means that it will
+	// reset the backoff state and cancel the pending backoff timer. Note that
+	// when cancelling the backoff timer, just like when the backoff timer fires
+	// normally, a new picker is returned to the channel, to force it to
+	// re-process any wait-for-ready RPCs that may still be queued if we failed
+	// them while we were in backoff. However, we should optimize this case by
+	// returning only one new picker, regardless of how many backoff timers are
+	// cancelled.
+
+	// Wait for the control channel to become READY.
+	for s := cc.cc.GetState(); s != connectivity.Ready; s = cc.cc.GetState() {
+		if !cc.cc.WaitForStateChange(cc.ctx, s) {
+			// cc.ctx has expired. Can happen only if close() was called.
+			return
+		}
+	}
+	cc.logger.Infof("Connectivity state is %s", cc.cc.GetState())
+
+	for {
+
+		// Wait for the control channel to enter anything other than READY.
+		state := cc.cc.GetState()
```

Comment:
Should this be `connectivity.Ready` directly?  What happens if it's not ready by the time it gets here due to a race?  Then we could stay stuck in another state forever.